hal/hal_uv_client_qemu.c

```diff
@@ -0,0 +1,227 @@
+    uv_close((uv_handle_t *)&client->pipe, qemu_uv_on_close);
+}
+
+static const adb_client_ops_t adb_qemu_uv_ops = {
+    .write = qemu_uv_write,
+    .kick  = qemu_uv_kick,
+    .close = qemu_uv_close
+};
+
+/* Please refer to:
+ * https://android.googlesource.com/platform/system/core/+/refs/heads/android11-dev/adb/daemon/transport_qemu.cpp#59
+ * https://android.googlesource.com/platform/system/core/+/refs/heads/android11-dev/qemu_pipe/qemu_pipe.cpp#37
+ * for qemu adb pipe protocol details.
+ */
+
+static void qemu_uv_on_setup(uv_handle_t *server) {
+    adb_context_uv_t *adbd = (adb_context_uv_t *)server->data;
+
+    adb_uv_qemu_setup(adbd);
+}
+
+static void qemu_uv_on_readable(uv_poll_t *server, int status, int events) {
+    int ret;
+    char buf[2];
+    uv_os_fd_t fd;
+    adb_client_qemu_t *client;
+    adb_context_uv_t *adbd = (adb_context_uv_t *)server->data;
+
+    ret = uv_fileno((uv_handle_t *)server, &fd);
+    assert(ret == 0);
+    uv_close((uv_handle_t *)server, qemu_uv_on_setup);
+
+    if (status < 0 || (events & UV_DISCONNECT)) {
+        adb_err("connect failed %d %d\n", status, events);
+        goto err;
+    }
+
+    ret = read(fd, buf, sizeof(buf));
+    if (ret != sizeof(buf)) {
+        adb_err("read failed %d %d\n", ret, errno);
+        goto err;
+    }
+
+    if (buf[0] != 'o' || buf[1] != 'k') {
+        adb_err("handshake failed\n");
+        goto err;
+    }
+
+    ret = write(fd, "start", 5);
+    if (ret != 5) {
+        adb_err("write failed %d %d\n", ret, errno);
+        goto err;
+    }
+
+    client = (adb_client_qemu_t *)adb_uv_create_client(sizeof(*client));
+    if (client == NULL) {
+        adb_err("failed to allocate client\n");
+        goto err;
+    }
+
+    /* Setup adb_client */
+    client->uc.client.ops = &adb_qemu_uv_ops;
+
+    ret = uv_pipe_init(adbd->loop, &client->pipe, 0);
+    /* TODO check return code */
+    assert(ret == 0);
+
+    ret = uv_pipe_open(&client->pipe, fd);
+    /* TODO check return code */
+    assert(ret == 0);
+
+    ret = uv_read_start((uv_stream_t *)&client->pipe,
+        qemu_uv_allocate_frame,
+        qemu_uv_on_data_available);
+    /* TODO check return code */
+    assert(ret == 0);
+    return;
+
+err:
+    close(fd);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+int adb_uv_qemu_setup(adb_context_uv_t *adbd) {
+    const char cookie[] = "pipe:qemud:adb:"
+                          STR(CONFIG_ADBD_QEMU_SERVER_PORT)
+                          "\0accept";
+    int ret;
+    int fd;
+
+    fd = open("/dev/goldfish_pipe", O_RDWR | O_CLOEXEC);
+    if (fd < 0) {
+        adb_err("qemu server open error %d %d\n", fd, errno);
+        return fd;
+    }
+
+    ret = write(fd, cookie, sizeof(cookie) - 1);
+    if (ret != sizeof(cookie) - 1) {
+        adb_err("qemu server write error %d %d\n", ret, errno);
```

Comment:
it's possible that ret is positive integer if fd is nonblocking (it should be with libuv).
https://man7.org/linux/man-pages/man2/write.2.html
> Note that a successful write() may transfer fewer than count bytes.

Reply:
Get, thank you.

Reply:
> it's possible that ret is positive integer if fd is nonblocking (it should be with libuv). https://man7.org/linux/man-pages/man2/write.2.html
> 
> > Note that a successful write() may transfer fewer than count bytes.

Link: https://github.com/spiriou/microADB/pull/47